### ISSUE TITLE
Add functionality for bears to exit cleanly 

### ISSUE THREAD

**Issue by abhsag24** [coala-analyzer/coala#2188]
right now there's no way a bear can exit cleanly after say logging an error, i think such type of a functionality would be nice to have for example in the `indentation bear` when there's a case of unmatched `{`, `}` .

**Comment by sils1297** [comment]
@abhsag24 then the indentation bear should ideally make a result explainig what's wrong right?

**Comment by abhsag24** [comment]
@sils1297 you're absolutely right :P , though i'm still leaving this issue open cause it's a feature worth having regardless :)

**Comment by sils1297** [comment]
can you file an issue against individual bears rather than coala then?

**Comment by abhsag24** [comment]
say we have a Bear of the form
```
class someBear:

    def run(self):
        some_function1()

    def some_function1():
        some_function2():

    def some_finction2():
        # code that requires this beear to exit
```
now if i want to exit from some_function2 i'd do a `self.err` and why the code failed, and then exit like a try-except block, with the functionality right now i'd have to go to rune by returning `None` in `some_function2` and `some_function1` and then `return []` in the `run` method.

**Comment by sils1297** [comment]
ah so something like a BearExit exception

**Comment by abhsag24** [comment]
^kind of but won't it give an ugly traceback?

**Comment by sils1297** [comment]
well coala would just catch it and print attache d info nicely if needed

Sincerely,

Lasse Schuirmann

lasse@schuirmann.net
http://coala-analyzer.org/ - http://viperdev.io/ - http://gitmate.com/

**Comment by abhsag24** [comment]
^that seems a like a good enough desing

**Comment by Makman2** [comment]
if you just want to exit while keeping control, raise a custom exception (call it `BearExit` or something else, like `UnmatchedParanthesesError`) and catch it inside `some_function1`, where you print the log and don't yield a result any more. I think we can avoid to define some kind of `BearExit`^^

**Comment by abhsag24** [comment]
@Makman2  i have done exactly that in my case :P , though wouldn't it be better for future Bears if coala catches this and we can just exit our bears anywhere by raising `BearExit` not sure ?

**Comment by Makman2** [comment]
I think there's no need to from coding perspective, in fact you can just do inside your run:
```
try:
    yield from code_that_runs()
except MyBearExit:
    return # or just pass as nothing follows
```

**Comment by AbdealiJK** [comment]
@Makman2 yes, but it may be better if we provide that boiler plate code to the bear writers.
The point is to make writing bears easy. I agree that this has not been required until now, but that's probably because we're focusing on linter bears which generally don't have lots of functions and are fairly small.
This may be needed for custom complicated bears.

If @abhsag24 's reason for creating this issue is over (he mentioned that in the second commit here) then we should close this for now until someone else needs something similar.

So, @abhsag24 do you think this will make your bear simpler? If not let's see if someone else needs it in the future and implement it when its needed

**Comment by abhsag24** [comment]
i am yielding a result when i raise the exception now, so it wont help my case anymore i kept this open since it'd be a nice little feature for future bears IMO, but we can close it, if it is not of use right now.,

**Comment by Makman2** [comment]
yeah let's do it, python has `SystemExit` which is quite similar, in fact we can even use this exception instead of having our own one^^
